File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/test/java/com/opspilot/TeamGraphTest.java
package com.opspilot;

import com.opspilot.team.BlackboardEntry;
import com.opspilot.team.RoleRunInput;
import com.opspilot.team.RoleRunResult;
import com.opspilot.team.RoleRunner;
import com.opspilot.team.SupervisorDecision;
import com.opspilot.team.TeamGraph;
import com.opspilot.team.TeamRole;
import com.opspilot.tools.Tool;
import org.junit.jupiter.api.Test;

import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static org.assertj.core.api.Assertions.assertThat;

class TeamGraphTest {

    private static RoleRunner fakeRunner(TeamRole role, BlackboardEntry.Kind kind, String content) {
        return new RoleRunner() {
            public TeamRole role() {
                return role;
            }

            public List<Tool> tools() {
                return List.of();
            }

            public RoleRunResult run(RoleRunInput input) {
                return new RoleRunResult(new BlackboardEntry(role, kind, input.brief(), content), List.of(), 1);
            }
        };
    }

    private static Map<TeamRole, RoleRunner> allRunners() {
        Map<TeamRole, RoleRunner> runners = new EnumMap<>(TeamRole.class);
        runners.put(TeamRole.ANALISTA, fakeRunner(TeamRole.ANALISTA, BlackboardEntry.Kind.FACTS, "fatos coletados"));
        runners.put(TeamRole.PLANEJADOR, fakeRunner(TeamRole.PLANEJADOR, BlackboardEntry.Kind.PLAN, "plano de 3 passos"));
        runners.put(TeamRole.EXECUTOR, fakeRunner(TeamRole.EXECUTOR, BlackboardEntry.Kind.EXECUTION, "acao executada"));
        return runners;
    }

    @Test
    void delegatesThenFinishesWithBriefAsAnswer() {
        TeamGraph graph = new TeamGraph(
                (message, blackboard, handoffCount) -> blackboard.isEmpty()
                        ? new SupervisorDecision(TeamRole.ANALISTA, "levante os fatos")
                        : SupervisorDecision.done("Resumo final do supervisor"),
                allRunners(), 1);

        TeamGraph.Result result = graph.run("o que esta acontecendo?");

        assertThat(result.answer()).isEqualTo("Resumo final do supervisor");
        assertThat(result.llmCalls()).isEqualTo(3); // 2 chamadas ao supervisor + 1 ao papel
    }

    @Test
    void fallsBackToBlackboardWhenDoneBriefIsEmpty() {
        AtomicInteger calls = new AtomicInteger();
        TeamGraph graph = new TeamGraph(
                (message, blackboard, handoffCount) -> {
                    if (calls.getAndIncrement() == 0) {
                        return new SupervisorDecision(TeamRole.ANALISTA, "levante os fatos");
                    }
                    return SupervisorDecision.done("");
                },
                allRunners(), 1);

        TeamGraph.Result result = graph.run("mensagem");

        assertThat(result.answer()).contains("Resumo do blackboard").contains("fatos coletados");
    }

    @Test
    void stopsAtHandoffCapEvenIfSupervisorNeverSaysDone() {
        TeamGraph graph = new TeamGraph(
                (message, blackboard, handoffCount) -> new SupervisorDecision(TeamRole.ANALISTA, "de novo"),
                allRunners(), 1);

        TeamGraph.Result result = graph.run("mensagem");

        long handoffEvents = result.trace().stream()
                .filter(event -> event.type() == com.opspilot.domain.TraceEvent.Type.HANDOFF)
                .count();
        assertThat(handoffEvents).isEqualTo(TeamGraph.MAX_HANDOFFS + 1); // +1 e o handoff final pra "done"
        assertThat(result.answer()).contains("Resumo do blackboard");
    }

    @Test
    void invalidSupervisorDecisionEndsGracefully() {
        TeamGraph graph = new TeamGraph(
                (message, blackboard, handoffCount) -> {
                    throw new IllegalArgumentException("json malformado");
                },
                allRunners(), 1);

        TeamGraph.Result result = graph.run("mensagem");

        assertThat(result.answer()).contains("A equipe encerrou sem contribuicoes");
        assertThat(result.trace().get(0).content()).contains(TeamGraph.INVALID_DECISION_PREFIX);
    }

    @Test
    void emptyMessageWithNoBlackboardStillProducesAnswer() {
        TeamGraph graph = new TeamGraph(
                (message, blackboard, handoffCount) -> SupervisorDecision.done(""),
                allRunners(), 1);

        TeamGraph.Result result = graph.run("mensagem vazia de contribuicoes");

        assertThat(result.answer()).contains("A equipe encerrou sem contribuicoes no blackboard para: mensagem vazia de contribuicoes");
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/team/team_graph_test.go
package team

import (
	"errors"
	"strings"
	"testing"

	"opspilot/domain"
	"opspilot/tools"
)

type fakeRunner struct {
	role    Role
	kind    BlackboardKind
	content string
}

func (r *fakeRunner) Role() Role          { return r.role }
func (r *fakeRunner) Tools() []tools.Tool { return nil }
func (r *fakeRunner) Run(input RoleRunInput) (RoleRunResult, error) {
	return RoleRunResult{
		Entry:    BlackboardEntry{Role: r.role, Kind: r.kind, Brief: input.Brief, Content: r.content},
		LLMCalls: 1,
	}, nil
}

func allRunners() map[Role]RoleRunner {
	return map[Role]RoleRunner{
		RoleAnalista:   &fakeRunner{role: RoleAnalista, kind: KindFacts, content: "fatos coletados"},
		RolePlanejador: &fakeRunner{role: RolePlanejador, kind: KindPlan, content: "plano de 3 passos"},
		RoleExecutor:   &fakeRunner{role: RoleExecutor, kind: KindExecution, content: "acao executada"},
	}
}

func TestDelegatesThenFinishesWithBriefAsAnswer(t *testing.T) {
	decideNext := func(message string, blackboard []BlackboardEntry, handoffCount int) (SupervisorDecision, error) {
		if len(blackboard) == 0 {
			return SupervisorDecision{Next: RoleAnalista, Brief: "levante os fatos"}, nil
		}
		return DoneDecision("Resumo final do supervisor"), nil
	}
	graph := NewTeamGraph(decideNext, allRunners(), 1)

	result := graph.Run("o que esta acontecendo?")

	if result.Answer != "Resumo final do supervisor" {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
	if result.LLMCalls != 3 { // 2 chamadas ao supervisor + 1 ao papel
		t.Fatalf("expected 3 llm calls, got %d", result.LLMCalls)
	}
}

func TestFallsBackToBlackboardWhenDoneBriefIsEmpty(t *testing.T) {
	calls := 0
	decideNext := func(message string, blackboard []BlackboardEntry, handoffCount int) (SupervisorDecision, error) {
		if calls == 0 {
			calls++
			return SupervisorDecision{Next: RoleAnalista, Brief: "levante os fatos"}, nil
		}
		return DoneDecision(""), nil
	}
	graph := NewTeamGraph(decideNext, allRunners(), 1)

	result := graph.Run("mensagem")

	if !strings.Contains(result.Answer, "Resumo do blackboard") || !strings.Contains(result.Answer, "fatos coletados") {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
}

func TestStopsAtHandoffCapEvenIfSupervisorNeverSaysDone(t *testing.T) {
	decideNext := func(message string, blackboard []BlackboardEntry, handoffCount int) (SupervisorDecision, error) {
		return SupervisorDecision{Next: RoleAnalista, Brief: "de novo"}, nil
	}
	graph := NewTeamGraph(decideNext, allRunners(), 1)

	result := graph.Run("mensagem")

	handoffEvents := 0
	for _, event := range result.Trace {
		if event.Type == domain.TraceHandoff {
			handoffEvents++
		}
	}
	if handoffEvents != MaxHandoffs+1 { // +1 é o handoff final pra "done"
		t.Fatalf("expected %d handoff events, got %d", MaxHandoffs+1, handoffEvents)
	}
	if !strings.Contains(result.Answer, "Resumo do blackboard") {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
}

func TestInvalidSupervisorDecisionEndsGracefully(t *testing.T) {
	decideNext := func(message string, blackboard []BlackboardEntry, handoffCount int) (SupervisorDecision, error) {
		return SupervisorDecision{}, errors.New("json malformado")
	}
	graph := NewTeamGraph(decideNext, allRunners(), 1)

	result := graph.Run("mensagem")

	if !strings.Contains(result.Answer, "A equipe encerrou sem contribuicoes") {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
	if !strings.Contains(result.Trace[0].Content, InvalidDecisionPrefix) {
		t.Fatalf("expected first trace event to contain invalid decision prefix, got %q", result.Trace[0].Content)
	}
}

func TestEmptyMessageWithNoBlackboardStillProducesAnswer(t *testing.T) {
	decideNext := func(message string, blackboard []BlackboardEntry, handoffCount int) (SupervisorDecision, error) {
		return DoneDecision(""), nil
	}
	graph := NewTeamGraph(decideNext, allRunners(), 1)

	result := graph.Run("mensagem vazia de contribuicoes")

	expected := "A equipe encerrou sem contribuicoes no blackboard para: mensagem vazia de contribuicoes"
	if !strings.Contains(result.Answer, expected) {
		t.Fatalf("unexpected answer: %q", result.Answer)
	}
}
